### ISSUE TITLE
How should copyto!(A, B) check bounds or shapes?

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/julia#26007]
Right now, `copyto!(dest, source)` uses two different bounds checking paradigms.  If the arrays are the same dimensionality, then it asserts that their shapes must be the same:

```
julia> copyto!([1 2 3], reshape(4:6, 1, 3))
1×3 Array{Int64,2}:
 4  5  6

julia> copyto!([1 2 3], reshape(4:6, 3, 1))
ERROR: BoundsError: attempt to access 1×3 Array{Int64,2} at index [Base.OneTo(3), Base.OneTo(1)]
Stacktrace:
 [1] throw_boundserror(::Array{Int64,2}, ::Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}) at ./abstractarray.jl:451
 [2] checkbounds at ./abstractarray.jl:416 [inlined]
 [3] copyto!(::Array{Int64,2}, ::Base.ReshapedArray{Int64,2,UnitRange{Int64},Tuple{}}) at ./multidimensional.jl:1153
 [4] top-level scope
```

But if they're different dimensionalities (or both `Array`), then `copyto!` only cares about the number of elements:

```
julia> copyto!([1 2 3], 4:6)
1×3 Array{Int64,2}:
 4  5  6

julia> copyto!([1 2 3], reshape([4,5,6], 3, 1))
1×3 Array{Int64,2}:
 4  5  6
```

We need to pick one, standardize, and document it.

**Comment by mbauman** [comment]
Given that this method also supports general iterables, I'd be inclined to just require that the number of elements match.  Moving in that direction is non-breaking.

**Comment by vchuravy** [comment]
Yeah the result of `copyto!(a::Array(2,3), b::Array(3,2))` should be the same as first reshaping `b` and than calling `copyto!`

**Comment by mbauman** [comment]
This is now fixed by #34049
